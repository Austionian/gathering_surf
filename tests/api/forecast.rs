use crate::{helpers::TestApp, mocked_happy_path_test_app, mocked_unhappy_path_test_app};

#[tokio::test]
async fn it_returns_the_forecast_data_as_json() {
    let app = mocked_happy_path_test_app!();

    let response = reqwest::get(format!("http://{}/api/forecast", &app.addr))
        .await
        .unwrap();

    assert_eq!(response.status().as_u16(), 200);

    let data = response.text().await.unwrap();

    assert!(data.contains("current_wave_direction"));
    assert!(data.contains("current_wave_height"));
    assert!(data.contains("forecast_as_of"));
    assert!(data.contains("qualities"));

    insta::assert_snapshot!(data);
}

#[tokio::test]
async fn it_handles_a_non_200_response_from_forecast_client_and_retries_once() {
    let app = mocked_unhappy_path_test_app!(forecast);

    let response = reqwest::get(format!("http://{}/api/forecast", &app.addr))
        .await
        .unwrap();

    assert_eq!(response.status().as_u16(), 500);

    let response = response.text().await.unwrap();

    assert!(!response.contains("current_wave_direction"));
    assert!(response.contains("Something went wrong: Non 200 response from NOAA"));
}

#[tokio::test]
async fn forecast_integration_test() {
    let app = TestApp::try_new()
        .await
        .expect("Unable to start test server.");

    let response = reqwest::get(format!("http://{}/api/forecast", &app.addr))
        .await
        .unwrap();

    assert_eq!(response.status().as_u16(), 200);

    let response = response.text().await.unwrap();

    assert!(response.contains("current_wave_direction"));
    assert!(response.contains("current_wave_height"));
    assert!(response.contains("forecast_as_of"));
    assert!(response.contains("qualities"));
}
