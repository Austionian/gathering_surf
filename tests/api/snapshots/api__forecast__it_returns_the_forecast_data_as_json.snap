---
source: tests/api/forecast.rs
assertion_line: 20
expression: data
---
{"forecast_as_of":"Mon, 10 Jun 2024 21:54:57","wave_height_data":[0.3048],"current_wave_height":"0","current_wave_direction":"","current_wave_period":"","wind_speed_data":[11.5],"wind_direction_data":[210.0],"wind_gust_data":[16.1],"wave_period_data":[4.0,4.0],"wave_height_labels":["Mon 03 PM"],"temperature":[60],"probability_of_precipitation":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"dewpoint":["44","44","44","44"],"cloud_cover":[4,4,4,4],"probability_of_thunder":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"starting_at":"2024-06-10T15:00:00-05:00","qualities":["#a8a29e"]}
